src/main/java/com/amazon/sqs/javamessaging/AmazonSQSExtendedClient.java

```diff
@@ -616,23 +621,36 @@ public SendMessageBatchResponse sendMessageBatch(SendMessageBatchRequest sendMes
             return super.sendMessageBatch(sendMessageBatchRequest);
         }
 
-        List<SendMessageBatchRequestEntry> batchEntries = new ArrayList<>(sendMessageBatchRequest.entries().size());
+        List<SendMessageBatchRequestEntry> originalEntries = sendMessageBatchRequest.entries();
+        ArrayList<SendMessageBatchRequestEntry> alteredEntries = new ArrayList<>(originalEntries.size());
+        alteredEntries.addAll(originalEntries);
 
+        // Batch entry sizes order by size
+        List<Pair<Integer, Long>> entrySizes = IntStream.range(0, originalEntries.size())
+                .boxed()
+                .map(i -> Pair.of(i, sizeOf(originalEntries.get(i))))
+                .sorted((p1, p2) -> Long.compare(p2.right(), p1.right()))
+                .collect(Collectors.toList());
+
+        long totalSize = entrySizes.stream().map(Pair::right).mapToLong(Long::longValue).sum();
+
+        // Move messages to s3 starting from the largest until total size is under the threshold if needed
         boolean hasS3Entries = false;
-        for (SendMessageBatchRequestEntry entry : sendMessageBatchRequest.entries()) {
-            //Check message attributes for ExtendedClient related constraints
-            checkMessageAttributes(clientConfiguration.getPayloadSizeThreshold(), entry.messageAttributes());
-
-            if (clientConfiguration.isAlwaysThroughS3()
-                || isLarge(clientConfiguration.getPayloadSizeThreshold(), entry)) {
-                entry = storeMessageInS3(entry);
-                hasS3Entries = true;
+        for (Pair<Integer, Long> pair : entrySizes) {
+            // Verify that total size of batch request is within limits
+            if (totalSize <= clientConfiguration.getPayloadSizeThreshold() && !clientConfiguration.isAlwaysThroughS3()) {
```

Comment:
Since we are checking the total size, does it make sense to move the if condition outside of the for loop? If it is true, we can throw an exception, else, we will keep add entry to s3

Reply:
@ziyanli-amazon the idea is to offload as few entries as possible, as each offload is a roundtrip against S3. So offloading goes from biggest entries to smallest until batch fits into size limit.